numpy/testing/_private/utils.py

```diff
@@ -821,7 +822,12 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
 
                     # note: this definition of relative error matches that one
                     # used by assert_allclose (found in np.isclose)
-                    max_rel_error = (error / abs(y)).max()
+                    # NaNs are disregarded to ensure that we report correct max
+                    # relative error between two arrays with zeros in the same position
+                    with warnings.catch_warnings():
```

Comment:
Fixing the error is nice. Do we have a way to avoid the warnings context here? If necessary by using numpy internals. Warning contexts are not thread-safe, and I think we had problems before with such things because some projects used threading during testing (and if just because they test their own threading support).

Reply:
We could use a stripped-down version of nanmax, but it would need to import ```_replace_nan``` from nanfunctions.py:
```python
rel_error = error / abs(y)
if type(rel_error) is ndarray and rel_error.dtype != object_:
    max_rel_error = fmax.reduce(rel_error, axis=None)
else:
    rel_error, mask = _replace_nan(rel_error, -inf)
    # If rel_error is full of NaNs print NaN as max relative error
    if mask is not None and all(mask):
        max_rel_error = array(nan)
    else:
        max_rel_error = amax(rel_error)
```
There don't seem to be any tests for ```assert_array_compare``` with object arrays though so I'm not sure if this solution is correct for them.

Reply:
@CakeWithSteak maybe you could add such a test (it should be pretty small after all?). Sorry if this is a bit annoying. We could make a small private helper (reusing `_replace_nan`) with a comment that the only goal is to not have a warning maybe.

Reply:
@seberg No problem! Is something like this ok?
```python
        def nanmax_nowarn(a, axis=None):
        """
        Same as nanmax, but doesn't raise a warning if the array is all NaN
        """
        from numpy import amax, fmax, all, object_
        from numpy.lib.nanfunctions import _replace_nan

        if type(a) is ndarray and a.dtype != object_:
            # Fast, but not safe for subclasses of ndarray, or object arrays,
            # which do not implement isnan (gh-9009), or fmax correctly (gh-8975)
            res = fmax.reduce(a, axis=axis)
        else:
            # Slow, but safe for subclasses of ndarray
            a, mask = _replace_nan(a, -inf)
            res = amax(a, axis=axis)
            if mask is None:
                return res

            # Check for all-NaN axis
            if mask.all():
                return array(nan)
        return res
```
One last question, hopefully: How should this handle subclasses of ```ndarray``` that don't implement ```all```? ```if mask.all()```would  raise ```NotImplementedError``` in cases like that.

Reply:
Those subclasses had issues before as well, no? The only would be duck types with `__array_function__` overrides, but those are brand new, and I not even sure if assert arrays compare doesn't convert to a numpy array by this time in any case.

I think this is right, the only thing I wonder if we can avoid the code duplication a bit....

Reply:
> I think this is right, the only thing I wonder if we can avoid the code duplication a bit....

Well the simplest way to avoid rewriting ```nanmax``` would be to add a new parameter to it which allows the caller to decide whether they want warnings or not. I'm not sure if you'd want to change the user facing API just for this fix though.

Reply:
Yeah, I am very much unwilling to do that. Might be willing to make a private helper. Although the indiscretion looks a bit weird for the fmax branch. OTOH, that is exactly why... if that branching logic needs fixing up, it will not happen...

Reply:
@CakeWithSteak @seberg is there any conclusion on a way forward here?

Reply:
@seberg By "private helper" do you mean a helper called by nanmax?
It's kind of messy because of the branching inside nanmax, this is what I've come up with:
```python
def _nanmax(a, axis=None, out=None, keepdims=np._NoValue):
    """
    Helper function for nanmax.

    Returns
    -------
    nanmax : ndarray
        An array with the same shape as `a`, with the specified axis removed.
        If `a` is a 0-d array, or if axis is None, an ndarray scalar is
        returned.  The same dtype as `a` is returned.
    warning_msg : {string, None}
        None if no warning should be raised, the warning message if an all-NaN slice is encountered.
    """
    kwargs = {}
    if keepdims is not np._NoValue:
        kwargs['keepdims'] = keepdims
    warning_msg = None
    if type(a) is np.ndarray and a.dtype != np.object_:
        # Fast, but not safe for subclasses of ndarray, or object arrays,
        # which do not implement isnan (gh-9009), or fmax correctly (gh-8975)
        res = np.fmax.reduce(a, axis=axis, out=out, **kwargs)
        if np.isnan(res).any():
            warning_msg = "All-NaN slice encountered"
    else:
        # Slow, but safe for subclasses of ndarray
        a, mask = _replace_nan(a, -np.inf)
        res = np.amax(a, axis=axis, out=out, **kwargs)
        if mask is None:
            return res, None

        # Check for all-NaN axis
        mask = np.all(mask, axis=axis, **kwargs)
        if np.any(mask):
            res = _copyto(res, np.nan, mask)
            warning_msg = "All-NaN axis encountered"
    return res, warning_msg


@array_function_dispatch(_nanmax_dispatcher)
def nanmax(a, axis=None, out=None, keepdims=np._NoValue):
    """..."""
    res, warning_msg = _nanmax(a, axis, out, keepdims)
    if warning_msg is not None:
        warnings.warn(warning_msg, RuntimeWarning,
                      stacklevel=3)
    return res
```
Then inside ```assert_array_compare``` we can do this:
```python
rel_error = error / abs(y)
max_rel_error, _ = _nanmax(rel_error)
```

Reply:
Yeah, I think this is probably the best solution here, thanks for the followup.

Reply:
Great! Should I change ```nanmin``` similarly for the sake of consistency or is that unnecessary?